SYCL/DeviceLib/string_test.cpp

```diff
@@ -0,0 +1,282 @@
+        s1_ptr[1] = s2_ptr[1] = 'x';
+        s1_ptr[2] = s2_ptr[2] = 'f';
+        s1_ptr[3] = s2_ptr[3] = '1';
+        s1_ptr[4] = s2_ptr[4] = '9';
+        s1_ptr[5] = s2_ptr[5] = 'T';
+        cmp_acc[0] = memcmp(s1_ptr, s2_ptr, 0);
+        cmp_acc[1] = memcmp(s1_ptr, s2_ptr, 1);
+        cmp_acc[2] = memcmp(s1_ptr, s2_ptr, 2);
+        cmp_acc[3] = memcmp(s1_ptr, s2_ptr, 3);
+        cmp_acc[4] = memcmp(s1_ptr, s2_ptr, 4);
+        cmp_acc[5] = memcmp(s1_ptr, s2_ptr, 5);
+        cmp_acc[6] = memcmp(s1_ptr, s2_ptr, 6);
+        s1_ptr[6] = 'Y';
+        s2_ptr[6] = 'U';
+        s1_ptr[7] = s2_ptr[7] = '7';
+        s1_ptr[8] = 'b';
+        s2_ptr[8] = 'g';
+        cmp_acc[7] = memcmp(s1_ptr, s2_ptr, 7);
+        cmp_acc[8] = memcmp(s1_ptr, s2_ptr, 8);
+        cmp_acc[9] = memcmp(s1_ptr, s2_ptr, 9);
+        s1_ptr[6] = 'U';
+        cmp_acc[10] = memcmp(s1_ptr, s2_ptr, 7);
+        cmp_acc[11] = memcmp(s1_ptr, s2_ptr, 8);
+        cmp_acc[12] = memcmp(s1_ptr, s2_ptr, 9);
+        s1_ptr[8] = 'g';
+        s1_ptr[9] = s2_ptr[9] = 'j';
+        cmp_acc[13] = memcmp(s1_ptr, s2_ptr, 10);
+        s1_ptr[10] = '1';
+        s2_ptr[10] = '2';
+        cmp_acc[14] = memcmp(s1_ptr, s2_ptr, 11);
+        s1_ptr[5] = 'Z';
+        cmp_acc[15] = memcmp(s1_ptr, s2_ptr, 11);
+      });
+    });
+  }
+
+  for (size_t idx = 0; idx < 13; ++idx) {
+    if ((cmps[idx] == 0 && refs[idx] == 0) || (cmps[idx] * refs[idx] > 0)) {
+      continue;
+    } else {
+      success = false;
+      break;
+    }
+  }
+
+  return success;
+}
+
+class KernelTestMemcpyAddrSpace;
+
+bool kernel_test_memcpy_addr_space(sycl::queue &deviceQueue) {
+  char src[16] = "abcdefg";
+  char dst[16];
+  char dst1[16];
+  {
+    sycl::buffer<char, 1> buffer1(src, sycl::range<1>(16));
+    sycl::buffer<char, 1> buffer2(dst, sycl::range<1>(16));
+    sycl::buffer<char, 1> buffer3(dst1, sycl::range<1>(16));
+    deviceQueue.submit([&](sycl::handler &cgh) {
+      sycl::accessor<char, 1, sycl::access::mode::read,
+                     sycl::access::target::constant_buffer,
+                     sycl::access::placeholder::false_t>
+          src_acc(buffer1, cgh);
+
+      sycl::accessor<char, 1, sycl::access::mode::read_write,
+                     sycl::access::target::local,
+                     sycl::access::placeholder::false_t>
+          local_acc(sycl::range<1>(16), cgh);
+
+      sycl::accessor<char, 1, sycl::access::mode::write,
+                     sycl::access::target::global_buffer,
+                     sycl::access::placeholder::false_t>
+          dst_acc(buffer2, cgh);
+
+      sycl::accessor<char, 1, sycl::access::mode::write,
+                     sycl::access::target::global_buffer,
+                     sycl::access::placeholder::false_t>
+          dst1_acc(buffer3, cgh);
+      cgh.single_task<class KernelTestMemcpyAddrSpace>([=]() {
+        // memcpy from constant buffer to local buffer
+        memcpy(local_acc.get_pointer(), src_acc.get_pointer(), 8);
```

Comment:
Hi, @vzakhari 
kernel_test_memcpy_addr_space is used to test memcpy whose input pointers have different address space. Could you help review if it meets our requirement?
Thanks very much.